### PR TITLE
write query parameters to swagger definition

### PR DESCRIPTION
**Pull request by rgarcia**
addresses #159

### CODE REVIEW

**Comment by achew22** [comment]
Rafael,

Thanks so much for sending in a patch! It looks like the build monster is unhappy because your change modifies the output for the example. grpc-gateway has the examples folder checked in and verifies that there is no difference between the checked in version and what comes out during CI. It looks like your change also needs to be rebased onto the latest master.

Thanks again!

**Comment by rgarcia** [comment]
@achew22 regenerated the examples and rebased. Tests are passing now.

File: protoc-gen-swagger/genswagger/template.go [hunk 1]
```diff
@@ -13,6 +13,49 @@ import (
 	"github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/descriptor"
 )
 
+// messageToQueryParameters converts a message to a list of swagger query parameters.
+func messageToQueryParameters(message *descriptor.Message, reg *descriptor.Registry, pathParams []descriptor.Parameter) ([]swaggerParameterObject, error) {
+	var parameters []swaggerParameterObject
```
Comment (yugui) [inline protoc-gen-swagger/genswagger/template.go]: It was even better if you make variable names a bit shorter.
https://github.com/golang/go/wiki/CodeReviewComments#variable-names

File: protoc-gen-swagger/genswagger/template.go [hunk 2]
```diff
@@ -13,6 +13,49 @@ import (
...
+	var addParameterWithPrefix func(string, *descriptor.Field) error
```
Comment (yugui) [inline protoc-gen-swagger/genswagger/template.go]:
```go
addParameterWithPrefix := func(prefix string, field *descriptor.Field) error {
```

BTW, why do you need to keep this function nested?
IIUC, you could simply define a function like
```go
func queryParams(prefix string, field *descriptor.Field) ([]swaggerParameterObject, error)
```

File: protoc-gen-swagger/genswagger/template.go [hunk 3]
```diff
@@ -13,6 +13,49 @@ import (
...
+			return nil
+		}
+
+		// nested type, recurse
+		msg, err := reg.LookupMsg("", field.GetTypeName())
+		if err != nil {
+			return fmt.Errorf("unknown message type %s", field.GetTypeName())
+		}
+		for _, nestedField := range msg.Fields {
+			if err := addParameterWithPrefix(field.GetName()+".", nestedField); err != nil {
```
Comment (yugui) [inline protoc-gen-swagger/genswagger/template.go]: Does it work fine with deeply nested fields?

Also, it is fine if it is out of scope for now. But what do you think about that a protobuf message has the same message type as a field?

**Comment by tmc** [comment]
@rgarcia can you rebase this again?

**Comment by rgarcia** [comment]
Hey, sorry. Probably won't be able to get back to this until early next week.

**Comment by sdemos** [comment]
What are the chances of this getting merged soon? This is a feature that we would definitely like to have.

File: protoc-gen-swagger/genswagger/template.go (hunk 3 above)
Comment (achew22) [inline protoc-gen-swagger/genswagger/template.go]: I'm also interested in the answer to this. What happens for deeply nested fields?

**Comment by achew22** [comment]
@rgarcia are you still interested in getting this merged?

**Comment by rgarcia** [comment]
@achew22 sorry, our team took a different direction, so I don't think I can do the follow-up to get this merged.

**Comment by achew22** [comment]
@rgarcia, would you object to me taking over this PR and applying the changes?

Note to people following along, if I haven't pushed a PR by Friday assume that I've fallen in a black hole and you can take a whack at it if you want

**Comment by rgarcia** [comment]
@achew22 go for it!

**Comment by Cedric-Venet** [comment]
What is missing for this to be merged? Just to answer the question on deeply nested?

**Comment by jhayotte** [comment]
Waiting this PR to be merged too! This feature is highly wanted!

**Comment by rilinor** [comment]
Is there any chance to merge this pull request?

**Comment by tmc** [comment]
closed by #297

**Pull request closed without merging**